src/set_relation/set_relation.cc

```diff
@@ -3762,4 +3762,200 @@ Relation* Relation::superAffineRelation(UFCallMap* ufcmap)
 }
 
 
+/*****************************************************************************/
+#pragma mark -
+/*************** VisitorsubNonAffineSet *****************************/
+//! Vistor Class used in SuperAffineSet
+//  Used in traversing a Set/Relation to replace UFCs with symbolic constants
+//  We will build up a nonAffineSet (or nonAffineRelation), term by term.
+//  And if a term is UFCall, we will convert it to symbolic constant,
+//  which is pre-computed and stored in the ufcmap
+class VisitorsubNonAffineSet : public Visitor {
+  private:
+         UFCallMap* ufcmap;
+         Set* nonAffineSet;
+         Relation* nonAffineRelation;
+         Conjunction* nonAffineConj;
+         std::list<Conjunction*> nonMAffineConj;
+         Exp* nonAffineExp;
+         bool visit;       // helps to know which Exp is UFC argument
+  public:
+         VisitorsubNonAffineSet(UFCallMap* imap){ufcmap = imap;}
+         // We do not change any type of Term except for VarTerm
+         void preVisitTerm(Term * t) {
+             if(visit){ nonAffineExp->addTerm( t->clone() ); }
+         }
+         void preVisitUFCallTerm(UFCallTerm * t);
+         void preVisitTupleVarTerm(TupleVarTerm * t){
+             if(visit){ nonAffineExp->addTerm( t->clone() ); }
+         }
+         void preVisitVarTerm(VarTerm * t);
+         void preVisitTupleExpTerm(TupleExpTerm * t){ 
+             if(visit){ nonAffineExp->addTerm( t->clone() ); }
+         }
+
+         void preVisitExp(iegenlib::Exp * e);
+         void postVisitExp(iegenlib::Exp * e);
+         void preVisitConjunction(iegenlib::Conjunction * c);
+         void postVisitConjunction(iegenlib::Conjunction * c);
+         void postVisitSet(iegenlib::Set * s);
+         void postVisitRelation(iegenlib::Relation * r);
+         Set* getSet(){ return nonAffineSet; }
+         Relation* getRelation(){ return nonAffineRelation; }
+};
+
+//! We iterate over terms in Exp, if the term is not a VarTerm
+//  then we just add it to our non-affine set.
+//  On the other hand, if a symbolic constants is in our ufcmap,
+//  we need to turn it into corresponding UFCalls to make an non-affine set.
+void VisitorsubNonAffineSet::preVisitVarTerm(VarTerm * t)
+{
+    if(!visit){
+        return;
+    }
+
+    VarTerm* cvt = (VarTerm*)(t->clone());
+    int cof = cvt->coefficient();
+    cvt->setCoefficient(1);
+    std::string symCons = cvt->symbol();
+    UFCallTerm* ufc = new UFCallTerm ( ufcmap->find( symCons ) );
+    UFCallTerm foo("foo", 0);
+
+    if( *ufc == foo )
+    {
+        nonAffineExp->addTerm( t->clone() );
+    }
+    else
+    {
+        ufc->setCoefficient(cof);
+        nonAffineExp->addTerm( ufc );
+    }
+    delete cvt;
+}
+
+void VisitorsubNonAffineSet::preVisitUFCallTerm(UFCallTerm * t)
```

Comment:
See the examples of throwing an assert_exception.  That is what you want to throw.  You do not want to catch it.